Initial commit of the Prometheus/OpenTelemetry compatibility spec (#25)
open-telemetry/wg-prometheus: specification.md

# OpenTelemetry/Prometheus Compatibility Specification

Status: [Experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/document-status.md)

## Abstract

OpenTelemetry is aiming to provide compatibility with
Prometheus and OpenMetrics. This document explains the
extent of the support. OpenTelemetry collector and libraries
will align with the compatibility requirements defined in
this spec.

> ## Goals

Reviewer: I think in one of our first meetings there was the idea of "To the OpenTelemetry collector an OpenTelemetry target and a prometheus instrumented target is indistinguishable."

Author: Added this.

* OpenTelemetry collector can be used as a drop-in replacement
  for Prometheus server to scrape and export metrics data.

> * OpenTelemetry collector should export OLTP-compatible Prometheus

Reviewer 2: ```suggestion
* OpenTelemetry collector should export OTLP-compatible Prometheus
```

Author: Done.

  time series to OpenTelemetry metrics exporters.

> * OpenTelemetry libraries should implement exporters to publish

Reviewer 3: Does this include in-process endpoints the enable Kubernetes to use OTel?

Author: Is there an example of an in-process endpoint you can share? I'm not very familiar with in-process Prometheus endpoints.

Reviewer 3: @brancz could you give context here? I don't want to say something subtly wrong.

Reviewer: Just that the Kubernetes processes (kubelet, apiserver, etc.) must have the `/metrics` endpoint themselves and they must work without external components. Kubernetes has stability guarantees on these endpoints as well as some of the metrics contained in them which is not something that can be broken.

Just a side note though, in the last sig-instrumentation meeting there were various voices that it may not be feasible at all for Kubernetes to move away from prometheus/client_golang as Kubernetes makes a lot of use of multiple registries for multiple /metrics endpoints and uses custom collectors all over the place.

Reviewer 4: So far I believe we've been discussing Prometheus compatibility at the server and data-model level, not in the client API (@reyang take note). OTel SDKs will certainly support a `/metrics` endpoint, and OTel has concepts that enable different registries (separate SDKs). I believe that "custom collector" in OTel client terms translates into "asynchronous instrument" in the OTel model API.

OTel has not stated a compatibility goal to offer drop-in compatible clients for metrics the way it did for OpenTracing in the tracing area; it would be an interesting exercise to try to take the OTel-Go Metrics SDK and implement the Prometheus client API including custom collectors. I think it's possible, but would be a serious effort, and there would still be fine-level distinctions worth pointing out (e.g., we have a difference in how histograms are synchronized presently in these two clients).

Reviewer: Yeah, to be honest, for Kubernetes I'm not even entirely sure it's worth the effort to create a compatibility API, if sig-instrumentation deems it worth it to move to otel then putting in the work of migrating is probably better anyways which you'd eventually want even with a go compatible API. If we have the "a prometheus and otel target are indistinguishable to the collector" characteristic, then Kubernetes is covered by otel either way, I think that's all people really care about most. I just wanted to share it.

Author: As you mentioned, the current focus and scope is the collector. Let's discuss the requirements from the SDKs more in detail when we start working on the Phase 2.

  Prometheus metrics.

> ## Differences and Limitations

Reviewer: How about metadata? (help, type, etc. which is experimentally available in prometheus remote-write)

Reviewer 4: The OTLP proto includes Help and Units. Type information is part of the data point, and we have a concept known as "Temporality" to distinguish Delta and Cumulative points. There are several open issues being discussed here, including one about how to represent the GaugeHistogram type, for example: https://github.com/open-telemetry/opentelemetry-proto/issues

  difference causes how the state is maintained throughout the
  collection including how it’s handled in the OpenTelemetry
  collector.
* **Cumulative vs delta**: OpenTelemetry supports delta temporality
  whereas Prometheus always expects absolute/cumulative values. This
  breaks some components that produce and communicate deltas where
  cumulatives cannot be rebuilt before being exported to Prometheus.
* **Histogram boundaries**: Prometheus histogram boundaries are by
  lower equal (le) while OpenTelemetry histogram boundaries are
  greater equal (ge). (This difference will be resolved via #18)
* **Semantic conventions**: OpenTelemetry predefines semantic
  conventions to collect additional metadata with telemetry data.
  Prometheus users don’t follow the same conventions and Prometheus
  client library provided data may lack the semantic conventions
  available in OpenTelemetry libraries.

## Compatibility Requirements

Given the number of fundamental design goals between OpenTelemetry
and Prometheus, our aim is to close the gaps where possible and
make the right compromises to meet the goals defined in this document.

> OpenTelemery and Prometheus won’t be fully compatible but we will

Reviewer: Can we be explicit about what is meant here?

Reviewer 4: I do not believe we should be making any compromises when we say that we support Prometheus. 

The proposal being built up here, is that an OpenTelemetry collector can eventually be configured with a number of pipeline stages, using receivers, processors, and exporters, to behave **exactly** as a Prometheus server would in addition to all of the other things it's capable of doing.

I understand there has been a mixed message.

For example, I know that some members of the OTel group have suggested we not support OpenMetrics export, which would mean not supporting the Prometheus federation design. We should be saying that we can do this, simply put, as it's not much different than supporting a Prometheus exporter inside the SDKs.

Reviewer 3: NB: Federation implies that data went through a Prometheus and received cleanups, sanity checking, etc. That will need an in-depth look on its own to make sure we don't accidentally get something wrong.

Reviewer 4: Our goal in working on specs like this is that we can replace "went through a Prometheus and received cleanups, sanity checking, etc" with a firm statement about what is required. I like how @rakyll wrote

> To the OpenTelemetry collector, an OpenTelemetry target and a Prometheus instrumented target is indistinguishable.
  
I'd add something high-level like, "To the Prometheus server, an OpenTelemetry collector and another Prometheus server are indistinguishable for establishing a federation."

enable important use cases to enable OpenTelemetry for Prometheus
users. The following sections summarizes the expectations from
the collector and the libraries.

### Collector

> * Collector will implement a Prometheus remote write exporter.

Reviewer 5: Is it the case the the collector will produce identical remote write messages to Prometheus & the Grafana Agent when scraping the same targets with the same config?

Is not, what will the differences be?

Author: Yes, the collector will produce the identical remote write messages. As a side note, it'd be great if Prometheus had some sort of a compliance testing.

  Publishing a pull-based metrics handler with all collected
  metrics is not a scalable approach.
* Collector will support scraping and ingesting cumulative metrics.
  Prometheus doesn’t support deltas and there are cases where
  rebuilding the cumulatives from deltas is not possible/easy.
* Collector will support all discovery and scraping configuration
  options in the Prometheus server. Collector will ignore the
  alerting rules.
* Collector will support exporting cumulative series to

>   OLTP-compatible exporters.

Reviewer 2: ```suggestion
  OTLP-compatible exporters.
```

Author: Done.

> * Collector will produce an “up” metric as a gauge that is compatible

Reviewer 3: This seems incompatible with Prometheus' defintion of "up", also see https://docs.google.com/document/d/1hn-u6WKLHxIsqYT1_u6eh94lyQeXrFaAouMshJcQFXs/edit#heading=h.k5ptdzrtfw6

Author: Our conversations stalled here until OTel introduces a concept for "up". It will be a topic to discuss after the initial GA of the metrics spec and we can revise this according to the plan.

  with the Prometheus server “up” metric. If scrape succeeds,
  it will produce 1 and 0 if it fails.
* Collector will produce "instance" and "job" labels similar
  to the Prometheus server.

> * If a target disappears from the scrape, the collector will

Reviewer 6: Is this different from the "up" metric?  I may be misremembering, but I thought the "up" metric is enough to determine staleness.

Reviewer 6: Thanks!

  write explicit staleness markers for the respective timeseries.
* Collector won’t assume any OpenTelemetry semantic conventions
  might be in place in the scraped data. Collector may decorate
  the samples with some semantic convention attributes available
  in the collector.

> * Collector will support WAL if it has access to persistent storage.

Reviewer 6: Do we consider the WAL an external API, or just an internal implementation detail?  I agree it solves a number of problems, but i'm not sure if it is required to be compatible with prometheus

Reviewer 3: If you're talking about Prometheus WAL, that would make remote write support easier to implement.

Author: This was about Prometheus WAL for the purposes of being able to reuse the Prometheus remote write. Today, we don't reuse the remote write from Prometheus and maintenance is a burden.

> * Collector will support WAL if it has access to persistent storage.

Author: Ack.

Status: [Experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/document-status.md)

## Abstract

OpenTelemetry is aiming to provide compatibility with
Prometheus and OpenMetrics. This document explains the
extent of the support. OpenTelemetry collector and libraries
will align with the compatibility requirements defined in
this spec.

## Goals

* OpenTelemetry collector can be used as a drop-in replacement
  for Prometheus server to scrape and export metrics data.
* OpenTelemetry collector should export OTLP-compatible Prometheus
  time series to OpenTelemetry metrics exporters.
* OpenTelemetry libraries should implement exporters to publish
  Prometheus metrics.
* To the OpenTelemetry collector, an OpenTelemetry target and
  a Prometheus instrumented target is indistinguishable.

## Differences and Limitations

OpenTelemetry and Prometheus/OpenMetrics have different design
goals and this reflects in the data model and implementation
details. This section summarizes a few key differences.

* **Pull vs push**: Prometheus is mainly designed for pull
  whereas OpenTelemetry primarily is designed for push. This

>   difference causes how the state is maintained throughout the
>   collection including how it’s handled in the OpenTelemetry

Reviewer 4: ```suggestion
  difference changes how state is maintained throughout the
  collection pipeline, including how it’s handled in the OpenTelemetry
```

  collector.
* **Cumulative vs delta**: OpenTelemetry supports delta temporality
  whereas Prometheus always expects absolute/cumulative values. This

>   breaks some components that produce and communicate deltas where

Reviewer 4: I'd like to be careful about what it means to "break" a component. The question, I guess, is what happens when Delta-temporality data arrives at a PRW or OpenMetrics exporter? That will indeed be an error, but that doesn't break Prometheus compatibility. The OTel data model is designed so that when used correctly (includes inside an SDK), Delta temporality can be translated to Cumulative temporality before one of these exporters sees the data.

Reviewer 3: Do deltas carry sequence numbers or similar? How does the rebuild deal with lost/overlapping/delayed data?

  cumulatives cannot be rebuilt before being exported to Prometheus.
* **Histogram boundaries**: Prometheus histogram boundaries are by
  lower equal (le) while OpenTelemetry histogram boundaries are

>   greater equal (ge). (This difference will be resolved via #18)

Reviewer 4: This merged and will go in the next release. https://github.com/open-telemetry/opentelemetry-proto/pull/262

* **Semantic conventions**: OpenTelemetry predefines semantic
  conventions to collect additional metadata with telemetry data.

>   Prometheus users don’t follow the same conventions and Prometheus

Reviewer 4: I believe we should be addressing this topic. It's not meant to create disruption for the Prometheus community, but the tracing and logging groups inside OpenTelemetry have been ironing out these semantic conventions for a while, with input from number of vendors, platform providers, and the Jaeger community, so we have a difficult to resolve situation. I believe we should be offering OTel collector pipeline stages to convert between these semantic conventions **in both directions**, especially for conventional metrics output by SDKs (e.g., runtime stats).

  client library provided data may lack the semantic conventions
  available in OpenTelemetry libraries.

## Compatibility Requirements

Given the number of fundamental design goals between OpenTelemetry
and Prometheus, our aim is to close the gaps where possible and
make the right compromises to meet the goals defined in this document.
OpenTelemery and Prometheus won’t be fully compatible but we will
enable important use cases to enable OpenTelemetry for Prometheus
users. The following sections summarizes the expectations from
the collector and the libraries.

### Collector

* Collector will implement a Prometheus remote write exporter.
  Publishing a pull-based metrics handler with all collected

>   metrics is not a scalable approach.

Reviewer 4: Please explain this sentence in more detail?

* Collector will support scraping and ingesting cumulative metrics.
  Prometheus doesn’t support deltas and there are cases where

>   rebuilding the cumulatives from deltas is not possible/easy.

Reviewer 4: Sure, but I wouldn't call that a Prometheus problem. It's on the user of deltas to be sure they are used correctly, and it's out of scope here.

Reviewer 3: I agree with Jaana's assessment that it's not possible to guarantee that no data is lost with delta, also see https://github.com/open-telemetry/wg-prometheus/pull/25#discussion_r595010522
That is one of the main reasons why Prometheus is designed for cumulatives: It avoids several complete classes of problems, at a cost to long-running processes with a high frequency of change.

This has been discussed in depth in 2017 and 2020, and we always came out at the same conclusion: You can always convert a cumulative into deltas, but not vice versa. At which point the question is what common case you optimize for.

* Collector will support all discovery and scraping configuration
  options in the Prometheus server. Collector will ignore the

>   alerting rules.

Reviewer 4: I agree about ignoring alerting rules, because it seems like a different reliability domain and often access to global state is required. We rarely expect to have global state in an OTel collector. This logic also applies to recording rules, but not so easily. Some recording rules can be implemented without access to global state, and those are ones we can eventually support inside OTel, as in the data model supports it but significant engineering work is needed.

* Collector will support exporting cumulative series to
  OTLP-compatible exporters.

> * At each scrape, collector will produce an "up" metric as a gauge

Reviewer 4: I will pursue the proposal discussed in this [thread](https://github.com/open-telemetry/opentelemetry-specification/issues/1078#issuecomment-776404440), where the receiver will attach both "alive" and "present" metrics. The PRW exporter would translate "alive" and "present" into "up" and this will be easy for the Prometheus-to-PRW code path because points will be aligned by the receiver. 

The reason for my proposal is that we can push OTLP data with just the "alive" metric, and join it with "present" from service discovery, producing aligned data that can be exported through PRW in the same form, yielding an "up" metric in PRW from pushed OTLP data combined with service discovery data.

Reviewer 3: How often will "present" be sent, and how do you ensure its timing is aligned with "alive"? I seem to remember a discussion about pitfalls and corner cases with a similar approach from some time back. I will try to find it and post if it's useful here.

  internally in the reciever and exporter will translate it to a
  Prometheus up metric in the export time.
* Collector will produce "instance" and "job" labels similar
  to the Prometheus server.
* If a target disappears from the scrape, the collector will

>   write explicit staleness markers for the respective timeseries.

Reviewer 4: In my proposal, the receiver would write "present" but not "alive" when the target disappears, which signifies to the PRW exporter that all timeseries are stale (and so that "up" becomes 0).

* Collector won’t assume any OpenTelemetry semantic conventions
  might be in place in the scraped data. Collector may decorate
  the samples with some semantic convention attributes available
  in the collector.

> * Collector will support Promeheus WAL if it has access

Reviewer 4: I see this as some kind of "durable queue" option. For a single collector (by some definition), a log file will do as in the Prometheus WAL. For a group of collectors, this might look like a Kafka cluster. There will be reasons for both, I expect we'll see.
